src/trainer_hill.c

```diff
@@ -0,0 +1,1121 @@
+	lsls r5, r0, 4\n\
+	subs r5, r0\n\
+	lsls r5, 3\n\
+	subs r5, r0\n\
+	lsls r5, 3\n\
+	adds r4, r5\n\
+	movs r0, 0xE8\n\
+	lsls r0, 2\n\
+	add r0, r12\n\
+	adds r0, r4\n\
+	ldrh r0, [r0]\n\
+	movs r4, 0xF\n\
+	subs r4, r1\n\
+	asrs r0, r4\n\
+	movs r4, 0x1\n\
+	ands r0, r4\n\
+	muls r2, r3\n\
+	adds r2, r1\n\
+	adds r2, r5\n\
+	movs r1, 0xA8\n\
+	lsls r1, 2\n\
+	add r1, r12\n\
+	adds r1, r2\n\
+	ldrb r1, [r1]\n\
+	movs r2, 0x80\n\
+	lsls r2, 2\n\
+	adds r3, r2, 0\n\
+	movs r2, 0xC0\n\
+	lsls r2, 6\n\
+	lsls r0, 10\n\
+	orrs r0, r2\n\
+	orrs r1, r3\n\
+	orrs r0, r1\n\
+	pop {r4,r5}\n\
+	pop {r1}\n\
+	bx r1\n\
+	.pool");
+}
+#endif // NONMATCHING
+
+void sub_81D5FB4(u16 *mapArg)
+{
+    s32 i, j;
+    u16 *src, *dst;
+    u8 mapId = GetCurrentTrainerHillMapId();
+
+    if (mapId == 6)
+    {
+        InitMapFromSavedGame();
+        return;
+    }
+
+    SetUpDataStruct();
+    if (mapId == 5)
+    {
+        InitMapFromSavedGame();
+        FreeDataStruct();
+        return;
+    }
+
+    mapId = GetFloorId();
+    src = gMapHeader.mapLayout->map;
+    gBackupMapLayout.map = mapArg;
+    gBackupMapLayout.width = 31;
+    gBackupMapLayout.height = 35;
+    dst = mapArg + 224;
+    for (i = 0; i < 5; i++)
+    {
+        for (j = 0; j < 16; j++)
+            dst[j] = src[j];
+        dst += 31;
+        src += 16;
+    }
+
+    for (i = 0; i < 16; i++)
+    {
+        for (j = 0; j < 16; j++)
+            dst[j] = sub_81D5F58(mapId, j, i, 0x10);
+        dst += 31;
+    }
+
+    mapheader_run_script_with_tag_x1();
+    FreeDataStruct();
+}
+
+bool32 InTrainerHill(void)
+{
+    bool32 ret;
+
+    if (gMapHeader.mapLayoutId == 0x19F
+            || gMapHeader.mapLayoutId == 0x1A0
+            || gMapHeader.mapLayoutId == 0x1A1
+            || gMapHeader.mapLayoutId == 0x1A2)
+        ret = TRUE;
+    else
+        ret = FALSE;
+
+    return ret;
+}
+
+u8 GetCurrentTrainerHillMapId(void)
+{
+    u8 ret;
+
+    if (gMapHeader.mapLayoutId == 0x19F)
+        ret = 1;
+    else if (gMapHeader.mapLayoutId == 0x1A0)
+        ret = 2;
+    else if (gMapHeader.mapLayoutId == 0x1A1)
+        ret = 3;
+    else if (gMapHeader.mapLayoutId == 0x1A2)
+        ret = 4;
+    else if (gMapHeader.mapLayoutId == 0x1A3)
+        ret = 5;
+    else if (gMapHeader.mapLayoutId == 0x19E)
+        ret = 6;
+    else
+        ret = 0;
+
+    return ret;
+}
+
+static bool32 sub_81D6100(void)
+{
+    bool32 ret;
+
+    if (gMapHeader.mapLayoutId == 0x1A3)
+        ret = TRUE;
+    else
+        ret = FALSE;
+
+    return ret;
+}
+
+const struct WarpEvent* sub_81D6120(void)
+{
+    const struct MapHeader *header = Overworld_GetMapHeaderByGroupAndId(MAP_GROUP(TRAINER_HILL_4F), MAP_NUM(TRAINER_HILL_4F));
+
+    return &header->events->warps[1];
+}
+
+const struct WarpEvent* sub_81D6134(u8 warpEventId)
+{
+    u8 id;
+    const struct MapHeader *header;
+
+    if (warpEventId == 1)
+        return &gMapHeader.events->warps[1];
+
+    id = sub_81D6490();
+    if ((u8)(id - 1) >= 4)
```

Comment:
Probably `id == 0 || id > 4`, or something similar (these casts and subtractions in a comparison usually are)

Reply:
Good call